src/libvirtApi/domain.js

```diff
@@ -506,66 +481,49 @@ export function domainDesktopConsole({
+    if (diskInactiveXML && persistent)
+        detachFlags |= Enum.VIR_DOMAIN_AFFECT_CONFIG;
+
+    await call(connectionName, vmPath, 'org.libvirt.Domain', 'DetachDevice', [diskXML, detachFlags], { timeout, type: 'su' });
 }
 
 // Cannot use virt-xml until https://github.com/virt-manager/virt-manager/issues/357 is fixed
-export function domainDetachHostDevice({ connectionName, vmId, live, dev }) {
+export async function domainDetachHostDevice({ connectionName, vmId, live, dev }) {
     const source = getHostDevSourceObject(dev);
     if (!source)
-        return Promise.reject(new Error(`domainDetachHostDevice: could not determine device's source identifier`));
-
-    const hostDevPromises = [];
+        throw new Error("domainDetachHostDevice: could not determine device's source identifier");
 
     // hostdev's <address bus=... device=...> may be different between live XML and offline XML (or it may be present in live XML but missing in offline XML)
     // therefore we need to call DetachDevice twice with different hostdevXMLs, once for live XML and once for offline XML
-    hostDevPromises.push(
-        call(connectionName, vmId, 'org.libvirt.Domain', 'GetXMLDesc', [Enum.VIR_DOMAIN_XML_INACTIVE], { timeout, type: 'u' })
-                .then(domInactiveXml => {
-                    const hostdevInactiveXML = getHostDevElemBySource(domInactiveXml[0], source);
-                    if (hostdevInactiveXML)
-                        return call(connectionName, vmId, 'org.libvirt.Domain', 'DetachDevice', [hostdevInactiveXML, Enum.VIR_DOMAIN_AFFECT_CONFIG], { timeout, type: 'su' });
-                })
-    );
+    const [domInactiveXml] = await call(connectionName, vmId, 'org.libvirt.Domain', 'GetXMLDesc', [Enum.VIR_DOMAIN_XML_INACTIVE], { timeout, type: 'u' });
+    const hostdevInactiveXML = getHostDevElemBySource(domInactiveXml, source);
+    if (hostdevInactiveXML)
+        await call(connectionName, vmId, 'org.libvirt.Domain', 'DetachDevice', [hostdevInactiveXML, Enum.VIR_DOMAIN_AFFECT_CONFIG], { timeout, type: 'su' });
 
     if (live) {
-        hostDevPromises.push(
-            call(connectionName, vmId, 'org.libvirt.Domain', 'GetXMLDesc', [0], { timeout, type: 'u' })
-                    .then(domXml => {
-                        const hostdevXML = getHostDevElemBySource(domXml[0], source);
-
-                        return call(connectionName, vmId, 'org.libvirt.Domain', 'DetachDevice', [hostdevXML, Enum.VIR_DOMAIN_AFFECT_LIVE], { timeout, type: 'su' });
-                    })
-        );
-    }
+        const [domXml] = call(connectionName, vmId, 'org.libvirt.Domain', 'GetXMLDesc', [0], { timeout, type: 'u' });
+        const hostdevXML = getHostDevElemBySource(domXml, source);
 
-    return Promise.all(hostDevPromises);
+        await call(connectionName, vmId, 'org.libvirt.Domain', 'DetachDevice', [hostdevXML, Enum.VIR_DOMAIN_AFFECT_LIVE], { timeout, type: 'su' });
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1397-20240122-152937-9774631b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)